src/decoder/stream.rs

```diff
@@ -213,7 +213,14 @@ impl JpegReader {
         jpeg_tables: Option<Arc<Vec<u8>>>,
     ) -> io::Result<JpegReader> {
         // Read jpeg image data
-        let mut segment = vec![0; length as usize];
+        let mut segment = Vec::new();
+        if segment.try_reserve_exact(length as usize).is_err() {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("Unable to allocate memory of {} size for segment.", length),
+            ));
+        }
```

Comment:
Thanks for identifying this. However, we don't want to check the global allocator running out of resources. On a lot of platforms this happens either not soon enough (i.e. Linux overcommit) or will still affect application stability regardless. That is, the checks of `try_reserve` are pretty much orthogonal to the DOS concerns from large images.

We'll want to check against and deduct from the configuration allocation limits instead: https://github.com/image-rs/image-tiff/blob/master/src/decoder/mod.rs#L303

You can figure out a way to make the attribute available as a reference and check against it here.

Reply:
I've looked more precisely at this case. And I found out, that [here](https://github.com/image-rs/image-tiff/blob/0da97709163fe08a093a00ce1e244ab301d82c02/src/decoder/image.rs#L517) we get the size that further would be used for memory allocation. Maybe it is a nice place to check Limints, but this code is located in expand_chunk method (Image struct). And Image doesn't now about Decoder's Limints.
So, to do any checks we need to change API. Honestly, I don't know if it worth it.

Reply:
The `expand_chunk` method is part of the internal API. I think it is absolutely reasonable to change the internal API to enable enforcement of limits